src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,441 @@
+        /// If the peer is behind/equal to our best height an empty array is sent back.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessGetHeadersAsync(INetworkPeer peer, GetHeadersPayload getHeadersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(getHeadersPayload), getHeadersPayload);
+
+            // Ignoring "getheaders" from peers because node is in initial block download unless the peer is whitelisted.
+            // We don't want to reveal our position in IBD which can be used by attacker. Also we don't won't to deliver peers any blocks
+            // because that will slow down our own syncing process.
+            if (this.initialBlockDownloadState.IsInitialBlockDownload() && !peer.Behavior<ConnectionManagerBehavior>().Whitelisted)
+            {
+                this.logger.LogTrace("(-)[IGNORE_ON_IBD]");
+                return;
+            }
+
+            ChainedHeader fork = this.chain.FindFork(getHeadersPayload.BlockLocators);
+
+            if (fork != null)
+            {
+                var headers = new HeadersPayload();
+
+                foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+                {
+                    headers.Headers.Add(header.Header);
+
+                    if ((header.HashBlock == getHeadersPayload.HashStop) || (headers.Headers.Count == MaxItemsPerHeadersMessage))
+                        break;
+                }
+
+                this.logger.LogTrace("{0} headers were selected for sending, last one is '{1}'.", headers.Headers.Count, headers.Headers.Last().GetHash());
+
+                try
+                {
+                    await peer.SendMessageAsync(headers).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("Unable to send headers message to peer '{0}'.", peer.RemoteSocketEndpoint);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Processes "headers" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="headersPayload">Payload of "headers" message to process.</param>
+        /// <remarks>
+        /// "headers" message is sent in response to "getheaders" message or it is solicited
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "headers" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessHeadersAsync(INetworkPeer peer, HeadersPayload headersPayload)
```

Comment:
@noescape00  as per the above comment I can't approve this method as it violates SRP. We always need to keep in mind how much a particular  piece of ocde is doing and in this case I can see:

1. If check to see if header count is 0 (can be moved to the caller)
2. If check to see if header count is less than MaxItemsPerHeadersMessage (can be moved to the caller)
3. Check to see if headers are in order (separate method, which should be unit tested)
4. Then a lock that does more checks and eventually, given the right scenario, resyncs. (separate method).

All of these needs to either be in separate methods or decisive logic (ifs) moved to the caller.

Reply:
The guards totally belong here, not to the caller.

Reply:
First part of the method consists of guards, it's just there are a lot of them but they totally belong here. 

And after that we have a method with a single responsibility so it doesn't violate SRP